File: Dynamic_Programming/718.Maximum-Length-of-Repeated-Subarray/718.Maximum-Length-of-Repeated-Subarray.ts
function findLength(nums1: number[], nums2: number[]): number {
  const m = nums1.length;
  const n = nums2.length;
  nums1.unshift(0);
  nums2.unshift(0);
  const dp = new Array(m + 1).fill(null).map(() => new Array(n + 1).fill(0));
  let ret = 0;

  for (let i = 1; i <= m; i++) {
    for (let j = 1; j <= n; j++) {
      dp[i][j] = nums1[i] === nums2[j] ? dp[i - 1][j - 1] + 1 : 0;
      ret = Math.max(ret, dp[i][j]);
    }
  }

  return ret;
}

File: Dynamic_Programming/718.Maximum-Length-of-Repeated-Subarray/718.Maximum-Length-of-Repeated-Subarray.py
# dp[i][j]: 以nums1 i 和 nums2 j结尾的最长的重复子数组
class Solution:
    def findLength(self, nums1: List[int], nums2: List[int]) -> int:
        result = 0
        dp = [[0] * (len(nums2) + 1) for _ in range(len(nums1) + 1)]

        for i in range(1, len(nums1) + 1):
            for j in range(1, len(nums2) + 1):
                if nums1[i - 1] == nums2[j - 1]:
                    dp[i][j] = dp[i - 1][j - 1] + 1
                result = max(result, dp[i][j])

        return result
